### ISSUE TITLE
Create a Firefox Quantum version of TabFern

### ISSUE THREAD

**Issue by cxw42** [cxw42/TabFern#100]
Mentioned in #82 and #99.  

 - [ ] Add Mozilla shim to the project
 - [ ] Convert the `chrome.*` calls to `browser.*` calls using the shim

**Comment by cxw42** [comment]
Requested by Ismo Räävi via support request 2018/04/15

**Comment by devinrhode2** [comment]
Looks like the steps in "Hacking on TabFern" for the switchover branch might now be right... not sure... Shouldn't a dev change to switchover branch? And branch off of switchover? https://github.com/cxw42/TabFern/wiki/Hacking-on-TabFern

Last but not least, I'm not sure how to use the built extension in firefox!

**Comment by cxw42** [comment]
@devinrhode2 Unfortunately, I haven't had a chance to work on it.  It does load as a temporary add-on without crashing, but I wouldn't consider it particularly usable :) .  I would be delighted to have some help!  Yes, if you do have any time, I would ask that you please work off `switchover` if you are able.  I am currently focused on #35 and would be delighted to include FF support in the next release as well!

Using the built extension: See [Loading temporary add-ons in Firefox](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Temporary_Installation_in_Firefox).  Pick `public/manifest.json` as the thing to load.  See the top comment (https://github.com/cxw42/TabFern/issues/100#issue-282712399) for open concerns I am aware of.  Thank you!

**Comment by devinrhode2** [comment]
I tried loading this, I have the extension icon in the top right in FF, but clicking it doesn't do anything :(

<img width="795" alt="screen shot 2018-12-26 at 11 31 44 am" src="https://user-images.githubusercontent.com/539816/50452803-ff0a6780-0901-11e9-96f1-28d0d0793a9d.png">

**Comment by devinrhode2** [comment]
(So I guess that's what the status on FF is :)

**Comment by cxw42** [comment]
Yep, those are normal (broadly defined :D).  Did you see the extension window pop up on its own?  Try double-clicking the icon and see if you get the window (double-click = summon to current browser window).

When I load the temporary extension, I get the same manifest errors, but the TF window opens anyway.

**Comment by devinrhode2** [comment]
Double clicking the icon still does not do anything for me, also tried
disabling all extensions except for stylus, restarting, and still nothing
for single/double click.

I am using FF 64.0 64-bit, on macOS High Sierra.

When I first load the extension in about:debugging, I see this:

[image: Screen Shot 2018-12-26 at 1.16.38 PM.png]

When I reload I see this:

[image: Screen Shot 2018-12-26 at 12.54.22 PM.png]
The last commit in my working directory (from you) - on the switchover
branch, is:
commit 2deb0684eb30d4f1d230d8fa79df04f073d3d5ed (HEAD -> firefox-work,
origin/switchover, origin/firefox-work, switchover)
Author: Chris White <cxwembedded@gmail.com>
Date:   Tue Dec 18 22:01:13 2018 -0500

    Remove S.COLLAPSE_ON_PARTIAL_WIN_CLOSE

    Don't use the option I started adding in
    25752b33290885b25a83bf84ef0b097fee176aea.  It would require winOnRemoved
    to receive state information about why the window was closed, and I
don't
    think this feature is sufficiently valuable to take the memory and
    complexity hit.  Let me know if you disagree!

    Also, improve the ergonomics of "collapse on window close".  If you hit
    Close on an already-closed window, collapse it.

    Mentioned in #35.


-Devin <http://devinrhode2.github.io>
<http://devinrhode2.github.io/>

**Comment by cxw42** [comment]
@devinrhode2 Unfortunately, the screenshots did not come through.  However, I added a fix for the manifest issues, in case that helps.  Instead of `npx brunch ...`, say `npm run ff`.  That will build a `public-ff` tree.  Load `public-ff/manifest.json` in Firefox and see if that helps any.  Thanks!

**Comment by devinrhode2** [comment]
Original should come through now:

When I first load the extension in about:debugging, I see this:
<img width="1012" alt="screen shot 2018-12-26 at 1 16 38 pm" src="https://user-images.githubusercontent.com/539816/50456333-3f291480-0919-11e9-8d61-7d9ad23bb618.png">

When I reload I see this:
<img width="1012" alt="screen shot 2018-12-26 at 12 54 22 pm" src="https://user-images.githubusercontent.com/539816/50456332-3f291480-0919-11e9-8c11-90f7e4c5b81a.png">

After pulling that from switchover I see:
First load:
<img width="1012" alt="screen shot 2018-12-26 at 2 59 14 pm" src="https://user-images.githubusercontent.com/539816/50457740-9c29c800-0923-11e9-967d-1271c76a5577.png">

Reload:
<img width="1012" alt="screen shot 2018-12-26 at 2 59 26 pm" src="https://user-images.githubusercontent.com/539816/50457743-a51a9980-0923-11e9-9acc-d6fd10f87605.png">

**Comment by cxw42** [comment]
Thanks!  The screenshots did come through.  I am on Windows, which is why I think I might be having trouble replicating.  I did try a fresh `npm install` with all my global modules uninstalled, still without successfully replicating.  I can think of three next steps to try:

1. If you have `brunch` installed globally, remove it --- I am using a tweaked version.

1. (See https://developer.mozilla.org/en-US/docs/Tools/Working_with_iframes if you're not familiar with this.)  In the developer console, select the middle iframe (`win/main.html`).  Then say `log.setLevel(log.levels.INFO)`.  That may provide more debug information.

1. Would you please send me a copy of your generated `public-ff/bg/background.js`?  I'd like to see if there are any differences.

Thank you for working with me on this!

**Comment by devinrhode2** [comment]
1. No global brunch
2. Not working, we'll need to get creative here I think. `log.setLevel` throws a ReferenceError on `log`
3. Background.js: https://gist.github.com/devinrhode2/86ae0e9e8629ae01ccb0adb142debba4 (this is the only script that appears when I click "edit as html" in the background page)

**Comment by devinrhode2** [comment]
Some screenshots coming up..

**Comment by devinrhode2** [comment]
Showed inspecting the location.href of the extension background page, and contrasted with mashable.com - in top right there's a strange little windowing type of button that's for selecting the frame. Not available for the background page. I am surprised that the background page uses iframes though... why is that? I figure the thing to do will be to set the sub-frames to have hijacked console methods that use `YourWIndow.sendMessage` to log console messages from top frame (only in FF) - or, maybe, we could eliminate usage of iframes.
<img width="1336" alt="screen shot 2018-12-26 at 7 46 31 pm" src="https://user-images.githubusercontent.com/539816/50462061-6fd47280-0948-11e9-82f5-04478321a42e.png">
<img width="1336" alt="screen shot 2018-12-26 at 7 46 56 pm" src="https://user-images.githubusercontent.com/539816/50462062-6fd47280-0948-11e9-9ca7-9ac73c000dc3.png">
<img width="1280" alt="screen shot 2018-12-26 at 7 38 19 pm" src="https://user-images.githubusercontent.com/539816/50462063-6fd47280-0948-11e9-9168-2f8eb2bbd461.png">

**Comment by devinrhode2** [comment]
wow removing those added bits got the extension to just popup when I reloaded the background page!

**Comment by cxw42** [comment]
Great news!  Now is the time for `log.setLevel`.

Would you please start a new topic branch from this point for Firefox?  I'm going to do the same for issue 35.  That way we won't step on each other's toes, and the eventual PRs will be cleaner.

If you find anything I haven't listed above, feel free to ping me here or open a separate issue as you see fit.

I am going to update the wiki with some more development info in the next few days.  Re. the event-order issue, for example, every tab has to be represented in both the multidex and the jstree.  The jstree had a holding-pen node where we can hang the tree node for a new tab before its window arrives.

**Comment by cxw42** [comment]
@devinrhode2 Is this a success report or a failure report? :D  
- Red X -> "actually, button press" message is normal.  
- I don't know what `h0-main.js` is, or what those errors are.  
- I do know that until the first item in my top comment is fixed, all the tabs will be off by one in any window created after the TF window opens :( .

**Comment by devinrhode2** [comment]
Would it be possible to export tabs from TabFern Chrome and import in TabFern Firefox?

**Comment by devinrhode2** [comment]
The h0-main.js errors were from other extensions
<img width="611" alt="screen shot 2018-12-27 at 12 59 14 pm" src="https://user-images.githubusercontent.com/539816/50491311-49114d00-09d7-11e9-853e-2f24f2261fb1.png">

**Comment by devinrhode2** [comment]
It's a success report

**Comment by cxw42** [comment]
Re. save data - yes - should be no problem.  The save-data file is just JSON.

Ideally, there should be no difference in functionality between Chrome and FF when we're done.  I certainly have not intentionally added any differences.

**Comment by devinrhode2** [comment]
Ok so I haven't switched over yet, figured out I can port active tabs by just bookmarking all tabs, but when I restarted firefox, I had some sort of "browser crashed, do you want to restore session?" from firefox, and saw this:

<img width="892" alt="screen shot 2018-12-28 at 8 43 06 am" src="https://user-images.githubusercontent.com/539816/50518934-ff2f7200-0a7d-11e9-94bf-1ef56f46b69c.png">

TabFern is not actually _installed_, since I'm just loading it from the directory, so probably there's some issue with Tab Center Redux

**Comment by cxw42** [comment]
Indeed, it may well be an interaction bug.  Not sure why the crash would have happened, though.  I just found [this](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Testing_persistent_and_restart_features) which looks like it has useful info (most notably, switch to [Developer Edition](https://www.mozilla.org/en-US/firefox/developer/) and set xpinstall.signatures.required to false) - I'll try it myself when I get the chance.

**Comment by cxw42** [comment]
OK - I tried `web-ext` without success.  However, what did work for me (with commit https://github.com/cxw42/TabFern/commit/72ffa2f83f237deee4fae8c22f0f3b84cc85f6a8 on `switchover`) is:
- Install Developer Edition
- Set my homepage in Dev. Edition to `about:debugging`
Now I can start up Dev Edition, hit "Load temporary," pick the manifest from `public-ff`, and work on it.
- After making changes, I do have to run `npm run ff` and then reload the extension manually in `about:debugging` :/ .

At one point, something messed up the permissions on `public-ff`.  However, exiting Dev Edition, removing the dir, and recreating seems to have worked fine.

**Comment by samir-s** [comment]
Hi, 
I'm reporting success on FF 66.0.3
The css is kinda weird, but as far as I've tested, it's working... 

Thanks god you did it !

I'd love to tip you, but didn't find anything for that on TabFern webpage...

**Comment by cxw42** [comment]
@samir-s Thank you!  I'll get right on that :D .

Please back up frequently!  I don't know what differences between Chrome and Firefox may be lying in wait to cause issues.  I look forward to any feedback you may have.

**Comment by cxw42** [comment]
@devinrhode2 re. "cannot find module process from /" --- I now have an Ubuntu 18.04 setup and am seeing the same.  For now I am going to do the manual hack you suggested (removing the lines) :) .  If I think of anything better I'll let you know.

**Comment by cxw42** [comment]
@devinrhode2 Would you please join me at #179?  Thanks!

**Comment by rugk** [comment]
I assume you already know https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/Porting_a_Google_Chrome_extension

**Comment by cxw42** [comment]
@rugk Yes, indeed, but thanks for checking!  Do you happen to have any bandwidth to help out? :D 

The differences TF runs into are largely undocumented assumptions.  E.g., in Chrome, a window always has at least one tab, but that's not the case in Vivaldi.  In Firefox, the extension can't manipulate any `about:debugging` tab, unlike in Chrome.

**Comment by rugk** [comment]
I may have a look. Though [seeing that it has already worked](https://github.com/cxw42/TabFern/issues/100#issuecomment-496267314) and I don't exactly know what is still TODO here.
I see these tick boxes in the OP, but do all of them need to be solved?

**Comment by cxw42** [comment]
@rugk Thanks for considering!  Yes, all of the check boxes at the top are things that need to be fixed before the general public can use TF on FF.  There may be others that will show up once those are done.

Sometime in the next few days I am going to make what is now the `issue35` branch the mainline, so that's the branch to look at if you're investigating.  The license will be clarified to CC-BY-SA 3.0+ in the near future, unless any of the existing contributors object.

**Comment by rugk** [comment]
CC licenses should not really be used for software. (they are intended for everything else) Rather use something like GPL instead then.
Also this trademark claim in the license file is a little strange. Did you really register a trademark?

**Comment by cxw42** [comment]
@rugk That discussion is at #145

**Comment by 7vq** [comment]
Just a note for anyone wondering how functional this extension currently is: I tested it for 5 minutes and frankly if Tree Style Tab interop was implemented I would use it as-is. The only bugs I've seen are two from the OP. Specifically:

- The "When creating a new window ..." bug: when you make a new window, that first tab is not seen by the extension. All later tabs seem to work perfectly. (And if you have multiple tabs in the window and just close and re-open the first tab, that first tab then seems to work perfectly.)
- The `about:debugging` etc bug: you can save a window that contains an `about` page tab, but trying to re-open such windows throws an error. You can just remove that tab from the saved window though.

<br>

I installed the extension as @cxw42 recommended earlier (downloading the "webstore" zip from TabFern's releases):

> Using the built extension: See [Loading temporary add-ons in Firefox](https://extensionworkshop.com/documentation/develop/temporary-installation-in-firefox/). Pick `public/manifest.json` as the thing to load.


&nbsp;
&nbsp;


@cxw42 I skimmed this issue and https://github.com/cxw42/TabFern/issues/179. Given that the extension seemed to work for me (I'm on macOS), should `Cannot find "process" from "/"` be checked off in the OP of this issue? Or does that bug only show up if you try to build from source, or something?

<br>

<sub>(Not sure what the "openWindowForURL" checkbox/issue in the OP refers to.)</sub>

**Comment by cxw42** [comment]
@7vq Thanks very much for trying it out, and for the update!  Glad to hear the good news --- I haven't run it on FF in a while.

What version of Firefox did you use, and what version+CPU of macOS?

Re. #179, building from source produces exactly what goes in the ZIP file.  Therefore, I think that box can be checked! :tada:

**Comment by cxw42** [comment]
@7vq 

> Tree Style Tab interop

Would you please open a new issue for that?  I'm curious to hear more about what you're looking for.

**Comment by 7vq** [comment]
> Thanks very much for trying it out, and for the update!

Of course. Thanks for making it!  


> What version of Firefox did you use, and what version+CPU of macOS?

Firefox 96.0.1, Big Sur, Intel  



> > Tree Style Tab interop
>
> Would you please open a new issue for that?

Definitely! Here: https://github.com/cxw42/TabFern/issues/286

**Comment by cxw42** [comment]
@iamashwin99 sorry for the delay in responding!  I am not currently able to devote time to the Firefox version, unfortunately :( .  However, PRs are certainly welcome, and I'm happy to answer questions or provide guidance.

**Comment by rugk** [comment]
Just FYI: [Firefox now also has native support for vertical tabs](https://support.mozilla.org/kb/use-sidebar-access-tools-and-vertical-tabs), it might not be the same and has the same features as TabFern, but it has quite many (even tab groups work now).

**Comment by cxw42** [comment]
Thanks!  Yep, the state of tab management has come a long way since TO and TF.  I'd like to think TF has been part of the inspiration behind some of those improvements :) 

(But I still use TF every day!)

**Comment by iamashwin99** [comment]
I think the two main feature I still miss is the ability to have one window control all the tabs ( so regardless of which window / desktop workspace I am in i can quickly navigate to any other tab). The other one is the ability to store the list of tabs without having the window open and also remember it across restarts and reboots. Does anyone have any idea how one could achieve those?
